common/src/main/java/com/revenuecat/purchases/common/networking/ETagManager.kt

```diff
@@ -0,0 +1,99 @@
+package com.revenuecat.purchases.common.networking
+
+import android.content.Context
+import android.content.SharedPreferences
+import org.json.JSONObject
+import java.net.HttpURLConnection
+
+private const val SERIALIZATION_NAME_ETAG = "eTag"
+private const val SERIALIZATION_NAME_HTTPRESULT = "httpResult"
+
+internal const val ETAG_HEADER_NAME = "X-RevenueCat-ETag"
+
+data class HTTPResultWithETag(
+    val eTag: String,
+    val httpResult: HTTPResult
+) {
+    fun serialize(): String {
+        return JSONObject().apply {
+            put(SERIALIZATION_NAME_ETAG, eTag)
+            put(SERIALIZATION_NAME_HTTPRESULT, httpResult.serialize())
+        }.toString()
+    }
+
+    companion object {
+        fun deserialize(serialized: String): HTTPResultWithETag {
+            val jsonObject = JSONObject(serialized)
+            val eTag = jsonObject.getString(SERIALIZATION_NAME_ETAG)
+            val serializedHTTPResult = jsonObject.getString(SERIALIZATION_NAME_HTTPRESULT)
+            return HTTPResultWithETag(eTag, HTTPResult.deserialize(serializedHTTPResult))
+        }
+    }
+}
+
+class ETagManager(
+    private val prefs: SharedPreferences
+) {
+
+    internal fun getETagHeader(
+        path: String,
+        refreshETag: Boolean = false
+    ): Map<String, String> {
+        val eTagHeader = ETAG_HEADER_NAME to if (refreshETag) "" else getETag(path)
+        return mapOf(eTagHeader)
+    }
+
+    internal fun shouldUseCachedVersion(responseCode: Int) = responseCode == RCHTTPStatusCodes.NOT_MODIFIED
+
+    internal fun getStoredResult(path: String): HTTPResult? {
+        val storedResult = getStoredResultSavedInSharedPreferences(path)
+        return storedResult?.httpResult
+    }
+
+    internal fun storeBackendResultIfNoError(
+        path: String,
+        resultFromBackend: HTTPResult,
+        eTagInResponse: String
+    ) {
+        val responseCode = resultFromBackend.responseCode
+        if (responseCode != RCHTTPStatusCodes.NOT_MODIFIED && responseCode < RCHTTPStatusCodes.ERROR) {
+            storeResult(path, resultFromBackend, eTagInResponse)
+        }
+    }
+
+    @Synchronized
+    private fun storeResult(
+        path: String,
+        result: HTTPResult,
+        eTag: String
+    ) {
+        val httpResultWithETag = HTTPResultWithETag(eTag, result)
+        prefs.edit().putString(path, httpResultWithETag.serialize()).apply()
+    }
+
+    private fun getStoredResultSavedInSharedPreferences(path: String): HTTPResultWithETag? {
+        val serializedHTTPResultWithETag = prefs.getString(path, null)
+        return serializedHTTPResultWithETag?.let {
+            HTTPResultWithETag.deserialize(it)
+        }
+    }
+
+    private fun getETag(path: String): String {
+        return getStoredResultSavedInSharedPreferences(path)?.eTag.orEmpty()
+    }
+
+    @Synchronized
+    internal fun clearCaches() {
+        prefs.edit().clear().apply()
+    }
```

Comment:
I think internal methods should go at the top of the file, or rather: public -> internal -> protected -> private, since it's more likely than someone reading this class would be interested in the public stuff first, then implementation details

Reply:
Yes, I missed this one :)

Although technically from the Kotlin Coding Conventions:

> Do not sort the method declarations alphabetically or by visibility, and do not separate regular methods from extension methods. Instead, put related stuff together, so that someone reading the class from top to bottom can follow the logic of what's happening. Choose an order (either higher-level stuff first, or vice versa) and stick to it.

But I don't agree with that and I agree with you lol

Reply:
oh, interesting, happy to go with that instead if it's the convention